instrumentation/google.golang.org/grpc/otelgrpc/grpctrace.go

```diff
@@ -56,6 +57,21 @@ func newConfig(opts []Option) *config {
 	return c
 }
 
+// Filter is a predicate used to determine whether a given gRPC request should
+// be traced. A Filter must return true if the request should be traced.
+type Filter func(ctx context.Context, fullMethodName string) bool
```

Comment:
Is there a more general interface that could be used here? See, e.g., the [`otelhttp.Filter`](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/master/instrumentation/net/http/otelhttp/common.go#L41), which accepts an `http.Request`  and thus enables a [wide range of filters](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/master/instrumentation/net/http/otelhttp/filters/filters.go) to be implemented.  Only being able to inspect the method name seems limiting.

Reply:
For sure, I just wasn't sure given the differences between unary/stream across the client & server how to come to something consistent.

[OpenTracing](https://github.com/grpc-ecosystem/go-grpc-middleware/blob/48900393c7f37e2136684cf7b0657940f001048f/tracing/opentracing/options.go#L22) currently just uses method name.

Reply:
I think `ctx context.Context, fullMethodName string` is sufficient considering that all UnaryServerInterceptor, UnaryClientInterceptor, StreamClientInterceptor and StreamServerInterceptor has both values in common.

Or to give more flexibility to users, we might go with approach like this:

```go
type Filter func(ctx context.Context, unaryInfo *grpc.UnaryServerInfo, steamInfo *grpc.StreamServerInfo) bool
```
then for UnaryServerInterceptor we can do like:
```go
for _, f := range cfg.Filters {
	if !f(ctx, info, &grpc.StreamServerInfo{}) {
		return handler(ctx, req)
	}
}
```
and for StreamServerInterceptor:
```go
for _, f := range cfg.Filters {
	if !f(ctx, &grpc.UnaryServerInfo{}, info) {
		return handler(srv, ss)
	}
}
```

With the above approach we only handle cases for server interceptors, because client interceptors did not have those values.
Should we have different interface for client interceptors? So there are `WithServerFilters` and `WithClientFilters`.
Considering that there are notable differences in `grpc.UnaryClientInterceptor` which is
```go
func(ctx context.Context,
	method string,
	req, reply interface{},
	cc *grpc.ClientConn,
	invoker grpc.UnaryInvoker,
	callOpts ...grpc.CallOption,
)
```
and `grpc.StreamClientInterceptor` which is
```go
func(ctx context.Context,
	desc *grpc.StreamDesc,
	cc *grpc.ClientConn,
	method string,
	streamer grpc.Streamer,
	callOpts ...grpc.CallOption,
)
```
now should we support all interesting values from both structs to our interface?

Reply:
Or maybe this can be a simpler alternative?
```go
type FilterInfo struct {
	Context	context.Context
	Method	string
	IsServer	bool
	IsClient	bool
	// more fields we can support here later
}

type Filter func(ctx context.Context, info FilterInfo) bool
```
We can wait-and-see to expose more information in `FilterInfo` later on.